### PR TITLE
armv7a/zynq7000: add new syspage implementation, timer, GIC and platform ctl

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

This PR is a part of changes made in [HBuczynski/syspage](https://github.com/phoenix-rtos/phoenix-rtos-kernel/commits/HBuczynski/syspage) branch. In order to provide better readability of review, the changes were separated into smaller pieces.

The following changes consist of:
 - cleaning `_init.S`
 - making a common `pmap` for `armv7al`
 - adding GIC
 - adding timer implementation
 - supporting platform ctl

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (imx6ull, zynq7000, qemu-system-aarch64).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/armv7a/zynq7000/_init.S [hunk 1]
```diff
@@ -161,6 +195,32 @@ clear_ttl1:
 	bl _cpy4                                              /* Fill TTL1 with 4 TTL2's addresses; pmap_common.kptab consists of 4 TTL2     */
 
 
+	/* Exceptions vectors and stack TTL1 entry
+	 * map 4MB V 0xffc00000 -> TTL2 of pmap_common.excptab */
+	ldr r0, =(ADDR_TTL1 + (0xffc << 2))                   /* Entry address: TTL1 base address + entry index (0xffc the last 4 entries in TTL1) * 4 bytes (entry size) */
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/_init.S:200]: Put comments within the code (i.e. comment above the line it describes). I see the point in making comment this way, but it causes lines to be way too long

File: hal/armv7a/zynq7000/interrupts.c [hunk 2]
```diff
@@ -19,52 +19,225 @@
...
 	spinlock_t spinlock[SIZE_INTERRUPTS];
 	intr_handler_t *handlers[SIZE_INTERRUPTS];
 	unsigned int counters[SIZE_INTERRUPTS];
-} interrupts;
+} interrupts_common;
 
 
-extern void _end(void);
+/* Required configuration for SPI (Shared Peripheral Interrupts IRQID[32:95]) */
+static const u8 spiConf[] = {/* IRQID: 32-39 */ rising_edge, rising_edge, high_lvl, high_lvl, reserved, high_lvl, high_lvl, high_lvl,
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/interrupts.c:54]: Start in the new line, so you don't have to indent this much

File: hal/armv7a/zynq7000/zynq.c [hunk 3]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+	*state = (*(zynq_common.slcr + slcr_aper_clk_ctrl) >> dev) & 0x1;
+	_zynq_slcrLock();
+
+	return 0;
+}
+
+
+static int _zynq_setDevClk(int dev, char divisor0, char divisor1, char srcsel, char clkact0, char clkact1)
+{
+	u32 id = 0;
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:123]:
```suggestion
	u32 id = 0;
	int err = 0;
```

File: hal/armv7a/zynq7000/zynq.c [hunk 4]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+			break;
+
+		default:
+			_zynq_slcrLock();
+			return -1;
+	}
+
+	_zynq_slcrLock();
+
+	return 0;
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:184]: More robust with one exit point
```suggestion
		default:
			err = -1;
			break;
	}

	_zynq_slcrLock();

	return err;
```

File: hal/armv7a/zynq7000/zynq.c [hunk 5]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+			}
+			break;
+
+		case pctl_mioclock:
+			if (data->action == pctl_set)
+				ret = _zynq_setMioClk(data->mioclock.mio, data->mioclock.ref0,
+					data->mioclock.mux0, data->mioclock.ref1, data->mioclock.mux1);
+			else if (data->action == pctl_get)
+				ret = _zynq_getMioClk(data->mioclock.mio, &data->mioclock.ref0,
+					&data->mioclock.mux0, &data->mioclock.ref1, &data->mioclock.mux1);
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:383]: platformctl_t should be packed - so you're taking address of packed struct member here

File: hal/armv7a/zynq7000/zynq.c [hunk 6]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+			break;
+
+		case pctl_devclock:
+			if (data->action == pctl_set)
+				ret = _zynq_setDevClk(data->devclock.dev, data->devclock.divisor0, data->devclock.divisor1,
+					data->devclock.srcsel, data->devclock.clkact0, data->devclock.clkact1);
+			else if (data->action == pctl_get)
+				ret = _zynq_getDevClk(data->devclock.dev, &data->devclock.divisor0, &data->devclock.divisor1,
+					&data->devclock.srcsel, &data->devclock.clkact0, &data->devclock.clkact1);
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:392]: As above - taking address of packed struct member here

File: hal/armv7a/zynq7000/zynq.c [hunk 7]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+			break;
+
+		case pctl_mio:
+			if (data->action == pctl_set)
+				ret = _zynq_setMIO(data->mio.pin, data->mio.disableRcvr, data->mio.pullup, data->mio.ioType, data->mio.speed, data->mio.l0,
+					data->mio.l1, data->mio.l2, data->mio.l3, data->mio.triEnable);
+			else if (data->action == pctl_get)
+				ret = _zynq_getMIO(data->mio.pin, &data->mio.disableRcvr, &data->mio.pullup, &data->mio.ioType, &data->mio.speed, &data->mio.l0,
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:400]: As above - taking address of packed struct member here

File: hal/armv7a/zynq7000/zynq.c [hunk 8]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+
+	return 0;
+}
+
+
+static int _zynq_getMIO(int pin, char *disableRcvr, char *pullup, char *ioType, char *speed, char *l0, char *l1, char *l2, char *l3, char *triEnable)
+{
+	u32 val;
+
+	if (pin > 53)
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:338]: Or make it unsigned
```suggestion
	if (pin < 0 || pin > 53)
```

File: hal/armv7a/zynq7000/zynq.c [hunk 9]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+
+	return -1;
+}
+
+
+static int _zynq_setMIO(int pin, char disableRcvr, char pullup, char ioType, char speed, char l0, char l1, char l2, char l3, char triEnable)
+{
+	u32 val = 0;
+
+	if (pin > 53)
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:319]: Or make it unsigned
```suggestion
	if (pin < 0 || pin > 53)
```

File: hal/armv7a/zynq7000/zynq.c [hunk 10]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+			return -1;
+	}
+
+	return 0;
+}
+
+
+static int _zynq_setMioClk(int mio, char ref0, char mux0, char ref1, char mux1)
+{
+	if (mio == pctl_ctrl_can_mioclk) {
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/zynq.c:284]: Hmm, what is the point of even massing `int mio` argument? I suppose other clock are TODO? In that case comment about that would be nice

File: hal/armv7a/zynq7000/timer.c [hunk 11]
```diff
@@ -17,34 +17,140 @@
...
+
+
+static void timer_setPrescaler(u32 freq)
+{
+	u32 ticks;
+	u32 prescaler;
+
+	ticks = TTC_SRC_CLK_CPU_1x / freq;
+
+	prescaler = 0;
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/timer.c:101]:
```suggestion
	u32 ticks = TTC_SRC_CLK_CPU_1x / freq;
	u32 prescaler = 0;
```

File: hal/armv7a/zynq7000/timer.c [hunk 12]
```diff
@@ -17,34 +17,140 @@
...
+	while ((ticks >= 0xffff) && (prescaler < 0x10)) {
+		prescaler++;
+		ticks /= 2;
+	}
+
+	if (prescaler) {
+		/* Enable and set prescaler */
+		prescaler--;
+		*(timer_common.ttc + clk_ctrl) |= (prescaler << 1);
+		*(timer_common.ttc + clk_ctrl) |= 0x1;
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/timer.c:111]: So if these bits not being zero won't mess our operations
```suggestion
		*(timer_common.ttc + clk_ctrl) = (*(timer_common.ttc + clk_ctrl) & ~0x1f) | (prescaler << 1);
		*(timer_common.ttc + clk_ctrl) |= 0x1;
```

File: hal/armv7a/zynq7000/interrupts.c [hunk 13]
```diff
@@ -19,52 +19,225 @@
...
+	*(interrupts_common.gic + ddcr) |= 0x3;
+
+	*(interrupts_common.gic + cicr) &= ~0x3;
+
+	/* Initialize CPU Interface of the gic
+	 * set the maximum priority mask */
+	*(interrupts_common.gic + cpmr) = (*(interrupts_common.gic + cpmr) & ~0x1f) | 0x1f;
 
+	/* EnableS = 1; EnableNS = 1; AckCtl = 1; FIQEn = 0 */
+	*(interrupts_common.gic + cicr) = (*(interrupts_common.gic + cicr) & ~0x7) | 0x7;
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/interrupts.c:242]: You're setting these bits anyway, no need to clear them

```suggestion
	*(interrupts_common.gic + cpmr) |= 0x1f;

	/* EnableS = 1; EnableNS = 1; AckCtl = 1; FIQEn = 0 */
	*(interrupts_common.gic + cicr) |= 0x7;
```

File: hal/armv7a/zynq7000/interrupts.c [hunk 14]
```diff
@@ -19,52 +19,225 @@
...
+	*(interrupts_common.gic + dicfr0 + (irqn >> 4)) = mask | ((conf & 0x3) << ((irqn & 0xf) << 1));
+}
+
+
+static void interrupts_setCPU(unsigned int irqn, u32 cpuID)
+{
+	u32 mask;
+
+	if (irqn < SPI_FIRST_IRQID || irqn >= SIZE_INTERRUPTS)
+		return;
```
Comment (agkaminski) [inline hal/armv7a/zynq7000/interrupts.c:131]: Some functions have this sanity check, some not. You're using this functions only within this file in a controlled manner, so I think there is no need for this extra defensiveness

Comment (agkaminski) [reply hal/armv7a/zynq7000/interrupts.c:131]: Applies to `interrupts_setConf` too

File: hal/armv7a/imx6ull/Makefile [hunk 15]
```diff
@@ -4,5 +4,5 @@
 # Copyright 2021 Phoenix Systems
 #
 
-OBJS += $(addprefix $(PREFIX_O)hal/armv7a/imx6ull/, _init.o console.o imx6ull.o pmap.o timer.o interrupts.o)
+OBJS += $(addprefix $(PREFIX_O)hal/armv7a/imx6ull/, _init.o console.o imx6ull.o timer.o interrupts.o)
```
Comment (kemonats) [inline hal/armv7a/imx6ull/Makefile:7]: To consider adding a dependency here:
```suggestion
OBJS += $(addprefix $(PREFIX_O)hal/armv7a/imx6ull/, _init.o console.o imx6ull.o timer.o interrupts.o)

$(PREFIX_O)hal/armv7a/imx6ull/_init.o: hal/armv7a/_hal.S hal/armv7a/_interrupts.S
```
or
```suggestion
OBJS += $(addprefix $(PREFIX_O)hal/armv7a/imx6ull/, _init.o console.o imx6ull.o timer.o interrupts.o)

$(PREFIX_O)hal/armv7a/imx6ull/_init.o: $(PREFIX_O)hal/armv7a/imx6ull/_init.d
```
same for `zynq7000`

File: hal/armv7a/zynq7000/zynq.c [hunk 16]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+{
+	u32 id = 0;
+
+	_zynq_slcrUnlock();
+	switch (dev) {
+		case pctl_ctrl_usb0_clk:
+		case pctl_ctrl_usb1_clk:
+			id = dev - pctl_ctrl_usb0_clk;
+			*(zynq_common.slcr + slcr_usb0_clk_ctrl + id) = (*(zynq_common.slcr + pctl_ctrl_usb0_clk + id) & ~0x00000070) | (srcsel & 0x7) << 4;
+			break;
```
Comment (kemonats) [inline hal/armv7a/zynq7000/zynq.c:131]: What do you think of this solution:
```suggestion
	volatile u32 *p;
	u32 tmp1, tmp2;

		case pctl_ctrl_usb0_clk:
		case pctl_ctrl_usb1_clk:
			id = dev - pctl_ctrl_usb0_clk;
			*p = zynq_common.slcr + slcr_usb0_clk_ctrl + id;
			tmp1 = *p & ~0x00000070;
			tmp2 = (srcsel & 0x7) << 4;
			break;
```
and after the `switch` statement
```
	*p = tmp1 | tmp2;
```

File: hal/armv7a/zynq7000/zynq.c [hunk 17]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+
+		case pctl_ctrl_gem0_rclk:
+		case pctl_ctrl_gem1_rclk:
+			id = dev - pctl_ctrl_gem0_rclk;
+			val = *(zynq_common.slcr + slcr_gem0_rclk_ctrl + id);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x1;
+			*clkact1 = *divisor0 = *divisor1 = 0;
+			return 0;
```
Comment (kemonats) [inline hal/armv7a/zynq7000/zynq.c:209]: perhaps
```suggestion
			break;
```

File: hal/armv7a/zynq7000/zynq.c [hunk 18]
```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
...
+		*ref1 = (val >> 22) & 0x1;
+
+		return 0;
+	}
+
+	return -1;
+}
+
+
+static int _zynq_setMIO(int pin, char disableRcvr, char pullup, char ioType, char speed, char l0, char l1, char l2, char l3, char triEnable)
```
Comment (kemonats) [inline hal/armv7a/zynq7000/zynq.c:315]: 10 members of the `mio` struct as arguments to the function?

File: hal/armv7a/zynq7000/zynq.c (hunk 5 above)
Comment (agkaminski) [reply hal/armv7a/zynq7000/zynq.c:383]: It's ok, it's a char - pointer can't be misaligned

File: hal/armv7a/zynq7000/zynq.c (hunk 17 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c]: Done, thanks.

File: hal/armv7a/zynq7000/interrupts.c (hunk 14 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/interrupts.c]: Done

File: hal/armv7a/zynq7000/interrupts.c (hunk 13 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/interrupts.c]: Done

File: hal/armv7a/zynq7000/timer.c (hunk 12 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/timer.c]: Done

File: hal/armv7a/zynq7000/timer.c (hunk 11 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/timer.c]: Done

File: hal/armv7a/zynq7000/zynq.c (hunk 9 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c:322]: I made it unsigned

File: hal/armv7a/zynq7000/zynq.c (hunk 8 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c:341]: I made it unsigned

File: hal/armv7a/zynq7000/zynq.c (hunk 4 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c]: Done

File: hal/armv7a/zynq7000/zynq.c (hunk 16 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c:132]: I don't want to add extra variables. I am leaving it as it is.

File: hal/armv7a/zynq7000/zynq.c (hunk 10 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c:287]: `MioClk` can be set only to `can`. To more readable I can add `err = -1` in else statement. What do you think?

File: hal/armv7a/zynq7000/_init.S (hunk 1 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/_init.S:200]: Hmm, my intention was to put the general comment above the code and describe each line of assembly code.
I would like to keep this format, my proposition it to break the comment line if it is too long.

File: hal/armv7a/zynq7000/zynq.c (hunk 10 above)
Comment (agkaminski) [reply hal/armv7a/zynq7000/zynq.c:287]: I think there is no point passing `mio` argument to this function, if only one special value is accepted. Better to call this (along with `platformctl`) along lines `setMioCanClk` and drop `mio` argument. Same applies to `_zynq_getMioClk`

File: hal/armv7a/zynq7000/zynq.c (hunk 18 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c]: Yeah, I know that it looks ugly. We use this approach to omit adding some extra structure definitions, it is internal function so it doesn't bother us so much.

File: hal/armv7a/zynq7000/_init.S (hunk 1 above)
Comment (agkaminski) [reply hal/armv7a/zynq7000/_init.S:200]: Look at e.g. line 135 - comment is much closer to the code. At least reduce number of spaces.

File: hal/armv7a/imx6ull/Makefile (hunk 15 above)
Comment (HBuczynski) [reply hal/armv7a/imx6ull/Makefile:7]: Done

File: hal/armv7a/zynq7000/zynq.c (hunk 10 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/zynq.c]: Done

File: hal/armv7a/zynq7000/_init.S (hunk 1 above)
Comment (HBuczynski) [reply hal/armv7a/zynq7000/_init.S]: Done

**Review by agkaminski — APPROVED** [review]

**Pull request merged**